### ISSUE TITLE
Lara jumps too far

### ISSUE THREAD

**Issue by vvs-** [opentomb/OpenTomb#38]
In *Caves* when Lara prepares to make a standing forward jump to the last secret from the ledge, she can only grab the opposite ledge to climb. But in OpenTomb she safely land behind the gap.

**Comment by Lwmte** [comment]
I think it is related to jump speed issue - in OpenTomb, Lara jumps significantly slower than in original. I will work on it after I finish hair and ragdoll functionality.

**Comment by vvs-** [comment]
`setEntityPos(player,77897,33218,-3072,83,0,0)` in *Caves*, then jump forward. Lara bumps her head against the rock. This doesn't happen in the original engine, i.e. she jumps too high.

**Comment by Joey79100** [comment]
It looks like many speeds are a bit "too much". If you're on the edge a block and you press backward, normally she should make a little hop and reach the opposite edge of the block, but instead her hop is a bit too long and she falls off it.
Same, when you have 2 blocks separated by a 2-block long pit, you can easily jump from one to another just by jumping from the edge of the first, while in the original games you need to press Action to grab the ledge, else you fall.

**Comment by vvs-** [comment]
Actually it's not just the constant. Modifying default speed makes Lara jump on a too short distance in one cases and still too far in others. Something is wrong in the whole formula.

**Comment by Lwmte** [comment]
Yes, it's completely wrong. Currently it is linear formula, i.e. each frame Lara is pushed at a constant height upwards, until affected by gravity. In original, jumps used either logarithmic or trigonometric formula, so Lara got larger momentum at the beginning of jump, and lower momentum at the end.

**Comment by stohrendorf** [comment]
A better approach would then be `speed[n+1] = speed[n] * (1-alpha*frameTime)`, where `alpha < 1` describes the damping/downwards accelleration per second.

**Comment by vvs-** [comment]
Can't the original computation be reverse engineered?

**Comment by Lwmte** [comment]
Yes. I tried to do it, and here is my result. Here's a snippet of jump code from TR1:

```c
  if (lara_pointer + 68) & 0x08 )  // 8 is a JUMP STATE flag.
  {
    lara_pointer_ = lara_pointer;
    
    // Some tricky business happens here...

    v70 = v67 + v68 * (*(_WORD *)(lara_pointer + 22) - 1 - *(_WORD *)(v81 + 16));
    *(_WORD *)(lara_pointer_ + 30) = *(_WORD *)(lara_pointer_ + 30) - HIWORD(v70) + ((unsigned int)(v68 + v70) >> 16);
    
    // Now, jump code itself.

    curr_vel = *(_WORD *)(lara_pointer_ + 32);   // That vertical velocity specified in AnimCommands
    gravity = 1;  // Simulates "air resistance"
    if ( curr_vel < 128 )    // 128 is a max. vertical velocity unaffected by "air resistance"
      gravity = 6;
    next_vel = curr_vel + gravity;
    *(_WORD *)(lara_pointer_ + 32) = next_vel;
    *(_DWORD *)(lara_pointer_ + 52) += next_vel;  // And what is this?!
  }
  else  // Move on floor?
  {
    if ( v68 )
      v67 += v68 * (*(_WORD *)(lara_pointer + 22) - *(_WORD *)(v81 + 16));
    *(_WORD *)(lara_pointer + 30) = HIWORD(v67);
    v69 = lara_pointer;
  }
```
As you can see, Lara's jumps in originals were very simple, without any extra "parabolic" calculations. I tried to do the same trick in `Character::freeFalling()` code by removing all fancy realistic physics calculations, but resulting jump trajectory was almost the same. I just can't understand how Core did this "bouncy" trajectory, when Lara moves upwards significantly faster than downwards.

**Comment by stohrendorf** [comment]
I played around with the constants, and this seems to work:
* Set `DEFAULT_CHARACTER_SLIDE_SPEED_MULT` to 80 instead of 75.
* Set `DEFAULT_CHARACTER_SPEED_MULT` to 30 instead of 31.5.

Otherwise changing the gravity to 1024*6 or 1000*6 (which looks more corresponding to that piece of decompilation) needs the default speed multiplier to stay over 30.5 to compensate large jump distance, which doesn't work on small jump distances, where the only working value is below 30.5. To compensate for a smaller default speed factor, the slide speed factor must of course be increased.

**Comment by vvs-** [comment]
No, these constants doesn't look authentic. They break a secret in *St. Francis' Folly* and the jumping distance there is shorter when she jumps to a lower slope and in *Palace Midas* it is longer when she jumps to a higher pillar. The sliding speed should be much lower than 75 otherwise Lara completely misses the slope in *St. Francis' Folly* and fall in the water instead.

**Comment by Lwmte** [comment]
Speed multiplier of 30 (why it was called like that in the first place? It's basically fixed-FPS multiplier value) is indeed authentic. In original TRs, with fixed FPS of 30, engine just added speed value to Lara's position each frame. So here we need to recalculate speeds based on variable FPS, and for this purpose we indeed need to multiply them by 30. I didn't check slide speed multiplier yet, so can't say anything about this.

Funny thing that while gravity value of 6000 and speed multiplier of 30 should produce correct jump heights, it does not. So either acceleration is applied incorrectly, or there is some other tricky business happening somewhere.

**Comment by vvs-** [comment]
I just don't see where the horizontal position is modified in that decompiled fragment. Seems that only vertical movement is there.

**Comment by Lwmte** [comment]
Horizontal position is completely other story. I can't find particular code for it in TR1 source (probably this tricky business two-liner does it), but in TR4 it's very clear that engine combines animation speed and AnimCommand speed in some manner. The process takes several steps and is much more complicated than vertical position modifier.

Probably the bad guy here is indeed horizontal position modifier, as it's directional jump animations which are not authentic. But still, it's not clear why speed multiplier and gravity values are similar yet producing different outcome in OpenTomb.

**Comment by vvs-** [comment]
I tested only directional jumps and sliding. I can't imagine how to test vertical jumps in isolation. There just seems to be no place depending on that in original levels.

**Comment by stohrendorf** [comment]
Assume we set gravity to 6000, and the multiplier to 30, then Lara's jumps are too short. Now when I think about jumping, I'll get some additional speed when jumping forward, meaning I'm faster than simply running. If it's done that way in the original engine, then it perfectly compensates the smaller multiplier and higher gravity.

**Comment by vvs-** [comment]
In the original she seems to be slower on the ascending trajectory and faster on the descending.

**Comment by T4Larson** [comment]
This is part of the corresponding movement code in TR4:
```c++
// pItem = &LaraObject
// pAnim->dFwdAccelPerFrame16_16 and pAnim->dFwdSpeed16_16 are 16.16 fixcomma values,
// hence the results are shifted >>16 to get the integer part.
if ( pItem->dFlags & ITF_IS_ABOVE_FLOOR )	// Lara in the air
{
	/************************************************************************************
	// "raw" code :
	//    somewhat awkwardly structured code,
	//    but essentially the same as the floor movement acceleration calc,
	//    except it doesn't check accel==0 (which doesn't affect the result):
	//
    animFwdSpeed = pAnim->dFwdSpeed16_16;
    tmpFwd =                  vAnimSpeedFrac + pAnim->dFwdAccelPerFrame16_16 * (pItem->wCurrentAnimFrame - pAnim->wFrameStart - 1);
    pItem->wSpeed_Forward -= (vAnimSpeedFrac + pAnim->dFwdAccelPerFrame16_16 * (pItem->wCurrentAnimFrame - pAnim->wFrameStart - 1)) >> 16;
    tmpSpeed = tmpFwd + pAnim->dFwdAccelPerFrame16_16;
    pItem->wSpeed_Forward +=  tmpSpeed >> 16;

	// --> same as:
	//    ...a bit shortened:
    tmpFwd = pAnim->dFwdSpeed16_16 + pAnim->dFwdAccelPerFrame16_16 * (pItem->wCurrentAnimFrame - pAnim->wFrameStart - 1);
    pItem->wSpeed_Forward -=  tmpFwd >> 16;
    pItem->wSpeed_Forward += (tmpFwd + pAnim->dFwdAccelPerFrame16_16) >> 16;

	************************************************************************************ */
	// --> same as:
	//    ...and turns out to be the same calculation as on floor:
    pItem->wSpeed_Forward = (pAnim->dFwdSpeed16_16 + pAnim->dFwdAccelPerFrame16_16 * (pItem->wCurrentAnimFrame - pAnim->wFrameStart)) >> 16;


	// --- calc and apply fall velocity:
    pItem->wSpeed_Y += pItem->wSpeed_Y >= 128 ? 1 : 6;
    pItem->dPos_Y += pItem->wSpeed_Y
}
else	// Lara on floor:
{
    animFwdAccel = pAnim->dFwdAccelPerFrame16_16;
    animFwdSpeed = pAnim->dFwdSpeed16_16;
    if ( animFwdAccel )
        animFwdSpeed += animFwdAccel * (pItem->wCurrentAnimFrame - pAnim->wFrameStart);
    pItem->wSpeed_Forward = animFwdSpeed >> 16;	// discard fraction
}

if ( dCurrentUsedRopeNr != -1 )
{
    calcRopeSwing(pItem);
}

if ( !(bLaraStatusBits_Lo & LST_BLOCK_MOVE_FORWARD) )
{
	// Apply forward velocity:
	// Note: wLaraFwdMoveAngle is a global value (tr-angle) 
	//       and may point to a different angle than Lara's rotation,
	//       i.e. sideways for side-jumps: this angle is adjusted
	//       as needed by the state-callbacks for an anim-state.
    fwdSpeed = pItem->wSpeed_Forward;
    pItem->dPos_X += fwdSpeed * 4 * tbl_Sinus[(wLaraFwdMoveAngle >> 3) & 0x1FFE] >> 14;
    pItem->dPos_Z += fwdSpeed * 4 * tbl_Cosinus[(wLaraFwdMoveAngle >> 3) & 0x1FFE] >> 14;
}
```
*[Sidenote: From what I can remember, for most animation the pAnim->dFwdAccelPerFrame16_16 is chosen so that the speed at the end of an animation matches the pItem->wSpeed_Forward of the next chained animation, i.e. the speed is linearl interpolated between the current animations movement speed the (chained) next animation: accel = (next_speed - current_speed) / num_anim_frames]*

With a bit of imagination, this TR4 code seems to be functionally identical to the TR1 code snippet posted by Lwmte above. It even has that awkwardly structured calculation in the "above floor" condition compared to the on-floor movement, and aparently hasn't changed much since TR1.

However, as you can see the velocity changes are controlled by the animation structure values for speed and acceleration, each frame.
This means doing simulationsteps asynchronously from animation-frame stepping may turn into a tricky nightmare to get the acceleration - and as such the precise movement distance - right.

Given the general precision problems with variable simulation timesteps, it may be more practical to use fixed timesteps for the physics (which isn't uncommon) and interpolate object movement for higher framerates...

**Comment by Lwmte** [comment]
Seems that really code is almost the same... Only difference is rope swing calculation - pretty funny that Core left all this code in the same place.

Regarding speed/accel values - we've sort of fixed it month or two ago, so there are no major problems with it. All we had to do is to multiply speed/accel by 30 (TR's own fixed timestep) and then just multipyly it by actual frame time. It produces fairly good result (possibly there are rounding errors, but they are really insignificant).

What is much more tricky is how engine processes AnimCommand 2 velocities here. Yes, vertical velocity seems to be linearly interpolated (although the curve in OpenTomb is still looks different, but I blame variable FPS and different camera for this), but what about *horizontal* velocity? This second value which "pushes" Lara forward in the air. It should be called `wSpeed_X` in your code, but there is no such value. Judging by TR4 disassembly, some tricky business goes on there - so possibly the value is treated just like `wSpeed_Forward`, and overrides animation's own speed? Or is it added to current animation speed?

Thank you for the snippet! :+1: May I ask, do you have any other TR4 code decompiled in the same manner? We can really learn from it. It seems much more valid than lots and lots of labels/comments spread around the assembly code in OllyDbg! :laughing:

**Comment by Lwmte** [comment]
Ok, from what I see in OllyDbg I can tell that "jump speed" overwrites animation speed. So technically, current OpenTomb speed calculation is valid. It's not clear why we get different curve though.

**Comment by Lwmte** [comment]
By the way, why do we have gravity value of ~6000 anyway? Shouldn't it be (`6 * TR_FRAME_RATE * frame_time)`?

**Comment by stohrendorf** [comment]
I guess you meant `6*TR_METERING_SIZE`

**Comment by Lwmte** [comment]
I also thought that gravity must be related to `TR_METERING_SIZE`, but what's the point? Gravity is completely independent on measurements or canonical dimensions, i. e. Core guys could set it to 5 or 7 instead of 6. Could there be any relation at all?

However, I have an idea. Let's assume that Core guys set gravity value of 6 by trial and error, i.e. 7 was too tight by the likes of Toby Gard / Paul Douglas / etc., and 5 was too weak. So they came up with value of 6, which precisely allows Lara to grab the ledge of a sector with a height of 1792 (7 clicks) on a given time, which we know is based on a 30 FPS timestep.

But if timestep in OpenTomb is (roughly) measured as 1/1000s of a second, what if we subtract additional `(6*1024) - (6*1000) = 144` from gravity value of 6000, resulting in value of 5856? Surprisingly, we get precise match of the jump height in this case, when combined with speed multiplier of 30.

I understand that my assumptions are absurd, since time and measurements have completely no relation here, but somehow this relation works.

**Comment by Lwmte** [comment]
P.S.: I'm only talking about jump heights here, of course - since this relation does not solve jump curve paradox.

**Comment by Lwmte** [comment]
Suddenly I've learned something else.
In originals, "normal" Lara jump-on-place height IS NOT equal to "reach" jump height. The best example of this is second secret in Caves. You can't grab the ledge with quick jump->action combination, you HAVE to use up+action combination instead: https://youtu.be/tScTbJW1n0E

Seems like whole jump code was strongly hardcoded by Core, completely ignoring animation properties and states. And I'm not yet talking about smash-jump cases, which are either dependent on frame numbers or vertical speed while jumping...

**Comment by stohrendorf** [comment]
Maybe the speed values are added, as running+jumping are basically two actions.

**Comment by Lwmte** [comment]
No, not possible - I checked it, speed is overwritten. If we add the speed, then Lara will jump at ridiculously far distance.

P.S.: Continue on reach jump height - yes, I found it. They overrided AnimCommand vertical speed for that particular "reach jump" case. It is on offset 4224F6 in TR4 (TRLE version). Speed is forced to -116. God, how I hate them... :imp: 

I see no reason to program such perverse workaround as well, so I'd recommend to overwrite AnimCommand value for animation 26 from -110 to -116.

**Comment by Lwmte** [comment]
I think we should implement some kind of script function to access level structures, because hardcoding all these overrides is no better than hardcoding speeds themselves.

**Comment by Lwmte** [comment]
Ok, latest commit moves this override to script. Although it's silly that we have to create new script function just for the sole purpose of modifying single parameter in a single animation.

**Comment by T4Larson** [comment]
> ...but what about horizontal velocity? This second value which "pushes" Lara forward in the air. It should be called wSpeed_X in your code, but there is no such value. Judging by TR4 disassembly, some tricky business goes on there - so possibly the value is treated just like wSpeed_Forward, and overrides animation's own speed? Or is it added to current animation speed?
All needed is _pItem->wSpeed_Forward_ and _wLaraFwdMoveAngle_, which defines the movement vector: See the end of the code snippet, where the speed vector is applied to _x_ and _z_ for **both** in-air and on-ground movement.

But: There's an error in the snippet I posted (this stuff is almost 10 years old... now I remember there was a reason why the in_air and on_ground calculations were slightl different), where the difference is that  in_air, the forward_speed is preserved and only acceleration is added - and this is also the reason, why in_air animations (jumping, falling, etc.) DON'T have speed and accel values (they're usually 0).
The correct arithmetic reduction would be:
```c++
	// --> same as:
	//    ...and turns out to be almost(!) the same calculation as on floor,
	//    CORRECTION: EXCEPT it does not set ABSOLUTE speed forward, but ADDS to forward speed:
	//    This means that once in the air, the momentum of pItem->wSpeed_Forward is preserved.
	//    (which makes sense when thought of as not having "ground friction" anymore).
    tmpAnimFwd = pAnim->dFwdSpeed16_16 + pAnim->dFwdAccelPerFrame16_16 * (pItem->wCurrentAnimFrame - pAnim->wFrameStart - 1);
    pItem->wSpeed_Forward += ((tmpAnimFwd + pAnim->dFwdAccelPerFrame16_16) >> 16) - (tmpAnimFwd >> 16);
```
Animcommand 2 parameters set **both** vertical (_pItem->wSpeed_Forward_) and horizontal (pItem->wSpeed_Y) values, this cmd also sets the ITF_IS_ABOVE_FLOOR flag. That's done in the intermediary animation right before the jump-fly-animation. After Animcmd 2 Lara is in in_air mode and uses the velocities from the AnimCmd 2.

**Comment by Lwmte** [comment]
The problem here is wall smash animations (32 and 33) and backwards-grab animation (94). They all have speed and/or accel specified in their fields. So they must be processed somehow.

I can't quite understand what you described, does it mean I need to add accel value to current speed, until speed specified in anim is reached?

**Comment by Lwmte** [comment]
Currently, acceleration and animation speed values are completely ignored for free falling move type. I asked TeslaRus to do something with this, but seems he wasn't able to do this.

I hope someone else will fix this situation.

**Comment by T4Larson** [comment]
> P.S.: Continue on reach jump height - yes, I found it. They overrided AnimCommand vertical speed for that particular "reach jump" case. It is on offset 4224F6 in TR4 (TRLE version). Speed is forced to -116. God, how I hate them... :imp:
>
I see no reason to program such perverse workaround as well, so I'd recommend to overwrite AnimCommand value for animation 26 from -110 to -116.

This "reach jump" does override the speed_y (vertical) once for the next AnimCmd2 that happens. But IIRC this is the only case where this happens for AnimCmd2.
Also, -116 is a case for when Lara can NOT reach the ledge, it only slightly differs from the speed_y value of -110 in the "jump-up" transition anim (animcmd2 in anim 91).
You probably noticed that Lara jumps up only so fast that she smoothly grabs the ledge at the apex of the jump, regardless of how hight the ledge is exactly:
There's a different calculation of the y-velocity for cases where she can reach such a ledge, wich must be at a height between 896 and 1920 tr-units:
```
wAnimCmd2SpeedYOverride = -3 - (int)sqrt(-9600 - 12 * -LedgeHeight);
```
Even if this is the only case where this happens for the AnimCmd2, there's still a lot of messing around with pLara->wSpeed_Y in the numerous state functions: Switching states or checking extra control keys when the velocity is in a certain range, or directly modifying the pLara->wSpeed_Y value when Lara hits the ceiling or is in grab-range  - not the same as animcmd2 override, but very frequent in the state functions.

A lot of the cases may be unimportant (such as small speedY adjustments to keep Lara swimming at the surface in the swim-state-callbacks) for gameplay, but more likely for jumping/flying there could be cases where it makes a difference to "just reach that ledge".

**Comment by Lwmte** [comment]
Huh, I play TRs for almost 15 years now, and never noticed this feature.
So it means we should override animation vertical speed *every time the state is called*. Cool! :sunglasses: 

And there are even more overrides? I got only two or three refs to that global override variable, all related to jumps. Or you mean underwater "push up" speed, which is activated on Lara's death?

**Comment by Lwmte** [comment]
I also don't see a point in "hard" override to -116 in "unreachable" case, cause formula calculates speed correctly in this case as well.

**Comment by Lwmte** [comment]
It's very sad, however, that we need to introduce such ugly magical overrides just to make the game look "authentic".

**Comment by T4Larson** [comment]
> The problem here is wall smash animations (32 and 33) and backwards-grab animation (94). They all have speed and/or accel specified in their fields. So they must be processed somehow.

Yes, these speed/accel values from the anim are used in the code which I posted - thats done for every tick (1/30s) - the only thing which is not set from the anim, is the _direction_ of that movement speed: The direction (_wLaraFwdMoveAngle_) is set from the state functions, i.e. the backwards-hop statefunction sets _wLaraFwdMoveAngle_ to -180deg from her looking orientation.
The animation still provides the movement speed, but _wLaraFwdMoveAngle_ points backwards and so during the hop she will be moved backwards.

**Comment by Lwmte** [comment]
The direction itself is indeed set in state control cases, in case of OpenTomb it's constants named `ENT_MOVE_FORWARD, ENT_MOVE_BACKWARD, ENT_MOVE_LEFT` and `ENT_MOVE_RIGHT`.

It's just my poor math knowledge which prevents me integrating speed calculations into free fall state code.

Funny that since TR4 they also made "lateral" speed/accel values, which are used for enemies, while they could simply implement same directional flags as for Lara and use them with normal animation speed/accel. The animation code of TRs is a pure mess from beginning (TR1) to end (TR5).

**Comment by vvs-** [comment]
Looks like after recent changes Lara can't reach some ledges, e.g. in *Atlantis*.

**Comment by Lwmte** [comment]
Can you provide some examples? Currently I can perfectly perform all the hard jumps.
Possibly my gravity calculation is wrong, and gravity should be lowered back to 5700 (although this value is a pure magic).

**Comment by vvs-** [comment]
Try a running jump at `68154,45984,-10240,0,0,0` or `54692,48205,-5120,180,0,0,0`. No matter what I do Lara can't grab the opposite ledge.

**Comment by Gh0stblade** [comment]
This is related to #28 :)

**Comment by Gh0stblade** [comment]
Now, some very interesting discussions here. Anyhow, before the other speed issues related to Lara's position updating are looked at. Y counterpart of the speed system needs to be fixed as some discussions earlier happened which I missed. Jump states are overrided it seems by the `character::jump` function. It needs to be modified so that the speed updates happen the same as the X counterpart per anim frame. I have some questions though, what about fall states? They don't seem to set the Y accel/Speed values accordingly, I'm guessing it was hardcoded in the original?

**Comment by Lwmte** [comment]
Yes, Y speeds were never set in animation code, everything was pretty much hardcoded (as described by T4Larson above).

``Character::jump`` doesn't override jump states, it only activates when corresponding AnimCommand is called. Its function is calculating vertical and horizontal speed, according to AnimCommand parameters.

However, I have noticed something. If you try to jump in any direction in slow-motion mode (press G key), Lara is pushed towards jump direction while not being in the air, i. e. in the end of "jump prepare" animations... Could it be the key of the problem?

**Comment by Lwmte** [comment]
I did additional checks, and indeed, "prepare to jump" animation speed (100) is completely equal to forthcoming vertical jump velocity (-100). Could it be that for jump prepare animations, engine used animation speed as a *vertical* speed?

**Comment by vvs-** [comment]
Unfortunately, in slow motion the engine works differently. I saw that on multiple occasions and don't use it for serious testing anymore.

**Comment by Gh0stblade** [comment]
@Lwmte Do you mean the 99 x speed? If so, that's correct, also happens in the original engine.

**Comment by vvs-** [comment]
No the frame rate is just calculated incorrectly. That's a cosmetic issue.

I've got completely different animations for issues #217 and #182 when tested in a slow motion.

**Comment by T4Larson** [comment]
This is a dump of a jump sequence in TR4:
```
***** Stand idle:
T:     0  /  State:   2  /  Anim: 103  /  SpdV:   0  /  SpdH:   0

***** Start Jump Forward:

T:    23  /  State:  15  /  Anim:  73  /  SpdV:   0  /  SpdH:   0
... 7 ticks unchanged ...
T:    31  /  State:   3  /  Anim:  76  /  SpdV:  99  /  SpdH:   0
... 2 ticks unchanged ...
T:    34  /  State:   3  /  Anim:  77  /  SpdV:  50  /  SpdH: -94
T:    35  /  State:   3  /  Anim:  77  /  SpdV:  50  /  SpdH: -88
T:    36  /  State:   3  /  Anim:  77  /  SpdV:  50  /  SpdH: -82
T:    37  /  State:   3  /  Anim:  77  /  SpdV:  50  /  SpdH: -76
...
... only speedh differs, gravity acceleration 6 units/tick^2
...
T:    62  /  State:   3  /  Anim:  77  /  SpdV:  50  /  SpdH:  74
T:    63  /  State:   3  /  Anim:  77  /  SpdV:  50  /  SpdH:  80
T:    64  /  State:   3  /  Anim:  77  /  SpdV:  50  /  SpdH:  86
T:    65  /  State:   3  /  Anim:  77  /  SpdV:  50  /  SpdH:  92
T:    66  /  State:   3  /  Anim:  77  /  SpdV:   0  /  SpdH:   0
T:    67  /  State:   2  /  Anim:  82  /  SpdV:   0  /  SpdH:   0
... 17 ticks unchanged ...
T:    85  /  State:   2  /  Anim:  11  /  SpdV:   0  /  SpdH:   0
... 1 ticks unchanged ...
T:    87  /  State:   2  /  Anim: 103  /  SpdV:   0  /  SpdH:   0
```
At Tick 31, SpeedV is taken from the Animation[76] speed/accel values (=99),
At the end of Anim 76, the Animcmd2 sets the values to 50,-100 which are carried over
to the next anim.
(Note that this dump is taken after frame start, so the speeds 50,-100 are effectively applied at the end of the previous frame, hence SpeedH in Frame 34 already has acceleration applied once).

**Comment by Gh0stblade** [comment]
There are quite a lot of *hacks* modifying the speed on specific animation states. Is it okay to rip it all out? The game seems to run fine without it, all speeds should be controlled by animation speed/accel values with some exceptions. I just feel the code is cluttering, confusing and redundant.

**Comment by vvs-** [comment]
Another problematic jump in *The Great Wall* at `51105,31251,1024,270,0,0`.

**Comment by vvs-** [comment]
In TR2 some places are easily accessible by jumping and climbing while they actually shouldn't. Another related issue is #346.

**Comment by vvs-** [comment]
I've started work on a test suite which should help to test and debug Lara trajectory problems. I'm not satisfied with the result yet, but very tired of experiments with TR engine. So, it's better to give a glimpse of what's inside in case somebody is curious enough. Right now there is no documentation, but I hope that it's self explanatory. This levels exposes many bugs in OT comparing to the original.

The file is here https://github.com/vvs-/testlevels/blob/master/trajectory1/LEVEL1.PHD

**Comment by vvs-** [comment]
BTW, it seems that its impossible or at least very hard to make a final jump in _Atlantean Stronghold_. Lara just can't reach the hole. Also, it's definitely impossible to make a similar jump in _Floating Islands_ in TR2. Well, there are plenty of places where the wrong jumping distance breaks puzzles making levels either ridiculously easy or impossible to complete.
